include/pybind11/pybind11.h

```diff
@@ -2194,24 +2199,44 @@ class class_ : public detail::generic_type {
     }
 
     /// Deallocates an instance; via holder, if constructed; otherwise via operator delete.
-    static void dealloc(detail::value_and_holder &v_h) {
+    static void dealloc_impl(detail::value_and_holder &v_h,
+                             bool release_gil_before_calling_cpp_dtor) {
         // We could be deallocating because we are cleaning up after a Python exception.
         // If so, the Python error indicator will be set. We need to clear that before
         // running the destructor, in case the destructor code calls more Python.
         // If we don't, the Python API will exit with an exception, and pybind11 will
         // throw error_already_set from the C++ destructor which is forbidden and triggers
         // std::terminate().
         error_scope scope;
-        if (v_h.holder_constructed()) {
-            v_h.holder<holder_type>().~holder_type();
-            v_h.set_holder_constructed(false);
-        } else {
-            detail::call_operator_delete(
-                v_h.value_ptr<type>(), v_h.type->type_size, v_h.type->type_align);
+        PyThreadState *py_ts = release_gil_before_calling_cpp_dtor ? PyEval_SaveThread() : nullptr;
```

Comment:
What about py::gil_scoped_acquire?

Reply:
Thanks for the nudge! This got me into some useful back and forth experimenting. The end result I settled on is functionally equivalent to what you saw before, but the code organization is better I think (avoids the `bool release_gil_before_calling_cpp_dtor`). I added comments to explain why I'm not using `gil_scoped_release`, and that the `catch (...)` isn't expected to be reachable.